### PR TITLE
[Example] NGCF

### PR DESCRIPTION
**Pull request by KounianhuaDu**
## Description
This PR is the implementation of the paper: Neural Graph Collaborative Filtering.

Current status:
1. On the gowalla dataset, when running 400 epochs and with the same settings, get a similar recall@20 result and a higher NDCG@20 result compared to the results reported in the paper.
2. The training process is slower than that of the original codes.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

**Comment by KounianhuaDu** [comment]
@zhjwy9343

**Comment by BarclayII** [comment]
Did you try profiling and see which statement/function/etc. is the bottleneck?  It's OK if you didn't.

**Comment by zhjwy9343** [comment]
In the layerwise profiling, computing in- and out-degree toke the major time.

In model level, forward consume the most and backward. 

So it seems the degree computation is the major bottleneck.

**Comment by KounianhuaDu** [comment]
@BarclayII  Yes, I've tried profiling. The profiling results are attached in the NGCF directory.  @zhjwy9343 Yes, it seems so. What if I store the degree information in ndata or edata beforehand? Or are there more efficient ways to implement the model? Thanks for your comments.

File: examples/pytorch/NGCF/README.md [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+The author's codes of implementation is in [here](https://github.com/xiangwang1223/neural_graph_collaborative_filtering). A pytorch re-implementation can be found [here](https://github.com/huangtinglin/NGCF-PyTorch).
+
+Example implementor
+----------------------
+This example was implemented by [Kounianhua Du](https://github.com/KounianhuaDu) during her Software Dev Engineer Intern work at the AWS Shanghai AI Lab.
+
+
+The graph dataset used in this example 
+---------------------------------------
+Gowalla: This is the check-in dataset obtained from Gowalla, where users share their locations by checking-in. To ensure thequality of the dataset, we use the 10-core setting, i.e., retainingusers and items with at least ten interactions
```
Comment (zhjwy9343) [inline examples/pytorch/NGCF/README.md:13]: thequality -> the quality, retrainingusers -> retaining users

File: examples/pytorch/NGCF/README.md [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+```
+
+NOTE: Following the paper's setting, the node dropout is disabled.
+
+
+Performance
+-------------------------
+The following results are the results in 400 epoches.
+**NGCF results**
+| Model         | Paper (tensorflow)               | ours (dgl)                  |
```
Comment (zhjwy9343) [inline examples/pytorch/NGCF/README.md:37]: dgl -> DGL

File: examples/pytorch/NGCF/NGCF/main.py [hunk 3]
```diff
@@ -0,0 +1,109 @@
+import torch
+import torch.optim as optim
+from model import NGCF
+from utility.batch_test import *
```
[the remark this answers is not in the document]
Comment (KounianhuaDu) [reply examples/pytorch/NGCF/NGCF/main.py:4]: Since there are lines that must be run in utility.batch_test before main, I have to use * here.

File: examples/pytorch/NGCF/NGCF/main.py [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+            mf_loss += batch_mf_loss
+            emb_loss += batch_emb_loss
+            
+
+        if (epoch + 1) % 10 != 0:
+            if args.verbose > 0 and epoch % args.verbose == 0:
+                perf_str = 'Epoch %d [%.1fs]: train==[%.5f=%.5f + %.5f]' % (
+                    epoch, time() - t1, loss, mf_loss, emb_loss)
+                print(perf_str)
+            continue
```
[the remark this answers is not in the document]
Comment (KounianhuaDu) [reply examples/pytorch/NGCF/NGCF/main.py:58]: Break the current epoch to miss the afterwards evaluation, which just need to be run every ten epochs. When "continue" is performed, it will end the current epoch and move to the next epoch.

File: examples/pytorch/NGCF/NGCF/main.py [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+
+        #evaluate the model every 10 epoches
```
[the remark this answers is not in the document]
Comment (KounianhuaDu) [reply examples/pytorch/NGCF/NGCF/main.py:60]: Since the "continue" line ends the current epoch. The evaluation will be run every 10 epochs.

**Comment by zhjwy9343** [comment]
LGTM

**Comment by yzh119** [comment]
@zhjwy9343 please notify contributors not to include data files in their PRs, which would make the repository bug and slow to clone.
Data should be compressed and uploaded to `data.dgl.ai`, and the repo should only keep data processing scripts.
We will revert this PR and cleanup the data files later.

**Comment by VoVAllen** [comment]
Reverted by https://github.com/dmlc/dgl/pull/2611

File: examples/pytorch/NGCF/README.md [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+
+Performance
+-------------------------
+The following results are the results in 400 epoches.
+
+**NGCF results**
+| Model         | Paper (tensorflow)               | ours (DGL)                  |
+| ------------- | -------------------------------- | --------------------------- |
+| recall@20     | 0.1569                           | 0.1552                      |
+| ndcg@20       | 0.1327                           | 0.2707                      |
```
Comment (yzh119) [inline examples/pytorch/NGCF/README.md:41]: The number is dubious, is `ndcg` the lower the better? If so we should figure out where the gap is.

File: examples/pytorch/NGCF/Data/gowalla/README.md [hunk 7]
```diff
@@ -0,0 +1,4 @@
+This is the check-in dataset obtained from Gowalla, where users share their locations by checking-in. 
+To ensure the quality of the dataset, the paper uses the 10-core setting, i.e., retaining users and items with at least ten interactions.
+Look for the full dataset?
+Please visit the [websit](https://snap.stanford.edu/data/loc-gowalla.html).
```
Comment (yzh119) [inline examples/pytorch/NGCF/Data/gowalla/README.md:4]: `websit`->`website`

**Review by yzh119 — COMMENTED** [review]
@KounianhuaDu would you mind creating another PR (we do not allow data files in PRs)? We appreciate your effort and thanks very much for your understanding.

Also, @zhjwy9343  I'm not sure if you profile the code in the correct way, PyTorch's execution is asynchronous on GPU and `time.time()` would not return you a correct measure. Would you mind elaborating more on the performance issue? I'll dig into details with the help of nsight-compute.

File: examples/pytorch/NGCF/README.md (hunk 6 above)
Comment (KounianhuaDu) [reply examples/pytorch/NGCF/README.md:41]: Running the author's code will get a similar result with ours. Both for our codes and the author's codes, the NDCG is just higher than the result reported in the paper.

Comment (KounianhuaDu) [reply examples/pytorch/NGCF/README.md:41]: It is the higher the better.

Comment (KounianhuaDu) [reply examples/pytorch/NGCF/README.md:41]: Another pr is launched. For the profile,  I set CUDA_LAUNCH_BLOCKING=1 . I'll attach the profile results in the new pr.
